source/aws_imds_client.c

```diff
@@ -0,0 +1,812 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/auth/aws_imds_client.h>
+#include <aws/common/clock.h>
+#include <aws/common/string.h>
+#include <aws/http/connection.h>
+#include <aws/http/request_response.h>
+#include <aws/http/status_code.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/socket.h>
+#include <ctype.h>
+
+#if defined(_MSC_VER)
+#    pragma warning(disable : 4204)
+#endif /* _MSC_VER */
+
+/* instance role credentials body response is currently ~ 1300 characters + name length */
+#define IMDS_RESPONSE_SIZE_INITIAL 2048
+#define IMDS_RESPONSE_TOKEN_SIZE_INITIAL 64
+#define IMDS_RESPONSE_SIZE_LIMIT 10000
+#define IMDS_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS 2
+#define IMDS_MAX_RETRIES 1
+
+struct aws_imds_client {
+    struct aws_allocator *allocator;
+    struct aws_http_connection_manager *connection_manager;
+    struct aws_retry_strategy *retry_strategy;
+    struct aws_imds_client_system_vtable *function_table;
+    struct aws_imds_client_shutdown_options shutdown_options;
+    /* will be set to true by default, means using IMDS V2 */
+    bool token_required;
+
+    struct aws_atomic_var ref_count;
+};
+
+static struct aws_imds_client_system_vtable s_default_function_table = {
+    .aws_http_connection_manager_new = aws_http_connection_manager_new,
+    .aws_http_connection_manager_release = aws_http_connection_manager_release,
+    .aws_http_connection_manager_acquire_connection = aws_http_connection_manager_acquire_connection,
+    .aws_http_connection_manager_release_connection = aws_http_connection_manager_release_connection,
+    .aws_http_connection_make_request = aws_http_connection_make_request,
+    .aws_http_stream_activate = aws_http_stream_activate,
+    .aws_http_stream_get_connection = aws_http_stream_get_connection,
+    .aws_http_stream_get_incoming_response_status = aws_http_stream_get_incoming_response_status,
+    .aws_http_stream_release = aws_http_stream_release,
+    .aws_http_connection_close = aws_http_connection_close,
+};
+
+static void s_aws_imds_client_destroy(struct aws_imds_client *client) {
+    if (!client) {
+        return;
+    }
+    aws_retry_strategy_release(client->retry_strategy);
```

Comment:
I feel like it's worth documenting that this function will only get called after all in-flight requests are finished and thus nothing is going to try and access the retry strategy after this point.